src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,141 @@
 package core.basesyntax;
 
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private static final Object[] EMPTY_LIST = {};
+
+    private Object[] listElements;
+    private int size = 0;
+
+    public ArrayList(int initialCapacity) {
+        if (initialCapacity > 0) {
+            this.listElements = new Object[initialCapacity];
+        } else if (initialCapacity == 0) {
+            this.listElements = EMPTY_LIST;
+        } else {
+            throw new IllegalArgumentException("Illegal Capacity: " + initialCapacity);
+        }
+    }
+
+    public ArrayList() {
+        this(DEFAULT_CAPACITY);
+    }
+
+    public ArrayList(Collection<? extends T> collection) {
+        this(collection.size());
+
+        for (T collectionElement : collection) {
+            add(collectionElement);
+        }
+    }
+
     @Override
     public void add(T value) {
+        ensureCapacity(size + 1);
 
+        this.listElements[size++] = value;
     }
 
     @Override
     public void add(T value, int index) {
+        ensureCapacity(size + 1);
+
+        size++;
+
+        checkIndex(index);
 
+        T[] afterPart = (T[]) Arrays.copyOfRange(listElements, index, size - 1);
+
+        this.listElements[index] = value;
+
+        for (int i = 0; i < afterPart.length; i++) {
+            this.listElements[index + 1 + i] = afterPart[i];
+        }
```

Comment:
Use [System.arraycopy()](https://docs.oracle.com/javase/8/docs/api/java/lang/System.html#arraycopy-java.lang.Object-int-java.lang.Object-int-int-) to move your array elements. Check all similar issues ;)